etw/native/src/main/cpp/etw_provider.cpp

```diff
@@ -0,0 +1,435 @@
+        } else {
+            jstring jstr_message = env->NewStringUTF(message);
+            rval = (jthrowable)env->NewObject(excls, init_id, jstr_message);
+            if (env->ExceptionCheck()) {
+                    DBG("Exception from constructor ApplicationInsightsEtwException.<init>(String)");
+                rval = NULL;
+            }
+            env->DeleteLocalRef(jstr_message);
+        }
+    }
+    env->DeleteLocalRef(excls);
+    return rval;
+}
+
+jthrowable newJniException(JNIEnv * env, const char * message, jthrowable cause) noexcept {
+    if (cause == NULL) {
+        return newJniException(env, message);
+    }
+
+    jthrowable rval = NULL;
+    jclass excls = env->FindClass("com/microsoft/applicationinsights/agentc/internal/diagnostics/etw/ApplicationInsightsEtwException");
+    if (excls == NULL) {
+        DBG("Could not find class ApplicationInsightsEtwException");
+        javaThrowUnknownError(env, " - could not find class ApplicationInsightsEtwException");
+    } else {
+        jmethodID init_id = env->GetMethodID(excls, "<init>", "(Ljava/lang/String;Ljava/lang/Throwable;)V");
+        if (init_id == NULL) {
+            DBG("Could not find constructor ApplicationInsightsEtwException.<init>(String, Throwable)");
+            javaThrowUnknownError(env, " - could not find constructor ApplicationInsightsEtwException.<init>(String, Throwable)");
+        } else {
+            jstring jstr_message = env->NewStringUTF(message);
+            rval = (jthrowable)env->NewObject(excls, init_id, jstr_message, cause);
+            if (env->ExceptionCheck()) {
+                    DBG("Exception from constructor ApplicationInsightsEtwException.<init>(String, Throwable)");
+                rval = NULL;
+            }
+            env->DeleteLocalRef(jstr_message);
+        }
+    }
+    env->DeleteLocalRef(excls);
+    return rval;
+}
+
+void javaThrowJniException(JNIEnv * env, std::string message) noexcept {
+    jthrowable cause = env->ExceptionOccurred(); // save the cause
+    jthrowable t;
+    env->ExceptionClear(); // for future exception check
+    if (cause) {
+        t = newJniException(env, message.c_str(), cause);
+    } else {
+        t = newJniException(env, message.c_str());
+    }
+
+    if (t) {
+        env->Throw(t);
+    } else {
+        if (!cause) {
+            javaThrowUnknownError(env, ": Error creating exception");
+        } else {
+            env->Throw(cause);
+        }
+    }
+    env->DeleteLocalRef(t);
+}
+
+void javaThrowUnknownError(JNIEnv * env, std::string message) noexcept {
+    jclass cls = env->FindClass("java/lang/RuntimeException");
+    if (cls != NULL) {
+        std::string m = "Unknown error from " DLL_FILENAME_STR + message;
+        env->ThrowNew(cls, m.c_str());
+    }
+    env->DeleteLocalRef(cls);
+}
+
+char * jstring2cstr(JNIEnv * env, jstring &jstr_input, char * cstr_output, aijnierr_t field_id) throw(aijnierr_t) {

```

Comment:
nit: since you're already using C++ **string** elsewhere - I would've just returned **std::string** from these helper methods. That will help you to get rid of a bunch of **delete[]** at the end of **writeEvent_IpaEtwEvent** function. Basically let the platform to manage the lifecycle of **std::string** for you, don't manage lifecycle of C-style const char * strings. Also your code would be smaller.

Reply:
Makes sense. I'll make this change.